File: app/src/main/java/com/worldventures/dreamtrips/modules/tripsimages/presenter/BaseImageViewPagerPresenter.java
package com.worldventures.dreamtrips.modules.tripsimages.presenter;

import com.worldventures.dreamtrips.modules.common.presenter.Presenter;
import com.worldventures.dreamtrips.modules.common.view.viewpager.FragmentItem;

import java.util.List;

import icepick.State;

public abstract class BaseImageViewPagerPresenter<VT extends BaseImageViewPagerPresenter.View> extends Presenter<VT> {
   private static final int LOAD_THRESHOLD = 5;

   protected boolean lastPageReached;
   protected boolean loading;

   @State int selectedPosition;

   public BaseImageViewPagerPresenter(boolean lastPageReached, int selectedPosition) {
      this.lastPageReached = lastPageReached;
      this.selectedPosition = selectedPosition;
   }

   @Override
   public void onViewTaken() {
      super.onViewTaken();
      initItems();
   }

   protected void initItems() {
      view.setItems(getItems());
      view.setSelectedPosition(selectedPosition);
   }

   protected abstract List<FragmentItem> getItems();

   public void pageSelected(int position) {
      if (!lastPageReached && !loading && getCurrentItemsSize() - position < LOAD_THRESHOLD)
         loadMore();
   }

   protected int getCurrentItemsSize() {
      return 0;
   }

   protected void loadMore() {

   }

   public interface View extends Presenter.View {
      void setSelectedPosition(int position);

      void setItems(List<FragmentItem> fragmentItems);
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/presenter/BaseImageViewPagerPresenter.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.presenter

import com.worldventures.dreamtrips.modules.common.presenter.Presenter
import com.worldventures.dreamtrips.modules.common.view.viewpager.FragmentItem
import icepick.State

abstract class BaseImageViewPagerPresenter<VT : BaseImageViewPagerPresenter.View>(
      internal var lastPageReached: Boolean,
      @JvmField @State var selectedPosition: Int) : Presenter<VT>() {

   internal var loading: Boolean = false

   protected abstract fun makeFragmentItems(): List<FragmentItem>

   open val currentItemsSize
      get() = 0

   override fun onViewTaken() {
      super.onViewTaken()
      initItems()
   }

   internal open fun initItems() {
      view.setItems(makeFragmentItems())
      view.setSelectedPosition(selectedPosition)
   }

   open fun pageSelected(position: Int) {
      if (!lastPageReached && !loading && currentItemsSize - position < VISIBLE_THRESHOLD) {
         loadMore()
      }
   }

   protected open fun loadMore() {}

   interface View : Presenter.View {
      fun setSelectedPosition(position: Int)

      fun setItems(fragmentItems: List<FragmentItem>)

      fun remove(position: Int)

      fun goBack()
   }

   companion object {
      private const val VISIBLE_THRESHOLD = 5
   }
}
